runtime/gc_vlhgc/InterRegionRememberedSet.cpp

```diff
@@ -935,6 +941,17 @@ MM_InterRegionRememberedSet::clearFromRegionReferencesForMarkDirect(MM_Environme
 					Card * cardAddress = rememberedSetCardToCardAddr(env, card);
 					/* Regions that are completely swept after a GMP, might still have outgoing references (thus we consider empty regions too) */
 					if (fromRegion->_markData._shouldMark  || !fromRegion->containsObjects() || isDirtyCardForPartialCollect(env, cardTable, cardAddress)) {
+						needToRemove = true;
+					}
+
+					if (!needToRemove && needToCheckMarkMap) {
```

Comment:
The new check is just more generic variant of fromRegion->containsObjects().
So I recommend you introduce a helper that will replace that containsObjects() check:

inline bool IRRS::cardContainsObjects(card, fromRegion,  needToCheckMarkMap) {
if needToCheckMarkMap
return markMap->anyLiveObjectInCard(card)
else 
return !fromRegion->containsObjects()
}

Reply:
We can consider checking against MarkMap all the time, if it's not much slower. But that might be tricky since for evacuated regions clearing of mark map might be done late(r).

Reply:
> The new check is just more generic variant of fromRegion->containsObjects().
> So I recommend you introduce a helper that will replace that containsObjects() check:
> 
> inline bool IRRS::cardContainsObjects(card, fromRegion, needToCheckMarkMap) {
> if needToCheckMarkMap
> return markMap->anyLiveObjectInCard(card)
> else
> return !fromRegion->containsObjects()
> }

good idea! it is good for clearFromRegionReferencesForMarkDirect(), but for clearFromRegionReferencesForMarkOptimized() (it is default, using compressed card table ), it is little bit tricky, partial logic for checking empty region is in rebuildCompressedCardTableForMark(), we can add CheckMarkMap in rebuildCompressedCardTableForMark, but it need to check every card in the regions, it might be expensive?

Reply:
I see. Actually, I think is CompressedCardTable::rebuildCompressedCardTableForPartialCollect that would be more impacted. We could try that later if it will slow it down.
Still I think you can introduce the helper and use it here, and in 'else' clause of 'if (compressedCardTable->isReady()) {', while the existing areAnyLiveObjectsInCard check you can move up in 'then' clause of 'if (compressedCardTable->isReady()) {'

Reply:
if we use the helper in  "else" clause, so we have to add the logic in rebuildCompressedCardTableForPartialCollect(), otherwise if compressedCardTable->isReady() case, we would miss to check markMap.
```
if (compressedCardTable->isReady()) {
   ....
} else {

}

```